man/portsdb.1
.\"
.Dd June 13, 2006
.Dt PORTSDB 1
.Os FreeBSD
.Sh NAME
.Nm portsdb ,
.Nm ports_glob
.Nd tools to manage and look up the ports database file
.Sh SYNOPSIS
.Nm
.Op Fl hfFquU
.Nm ports_glob
.Op Fl hMrR
.Op Fl x Ar pkgname_glob | portorigin_glob
.Op Ar pkgname_glob | portorigin_glob
.Ar ...
.Sh DESCRIPTION
The
.Nm
command is a tool to generates the ports database named
.Pa INDEX.db
from the ports index file named
.Pa INDEX .
It is commonly used among the tool suite and automatically updated on
demand when it gets older than the index file.
To save time, you can
update it beforehand, like right after the index file is updated.
Note that
.Pa INDEX
file is updated every few hours on official site,
it is recommended that you run
.Dq Nm portsdb Fl Fu
after every CVSup of the ports tree in order to keep them always
up-to-date and in sync with the ports tree.
.Pp
.Nm ports_glob
looks up the ports database to expand given patterns.
.Pp
Actually,
.Nm
and
.Nm ports_glob
share an entity and are totally equivalent.
.Pp
Before reading these instructions, you must understand that a
port/package can have the following two types of related
ports/packages:
.Bl -tag -width "dependent" -compact
.It required
Ports/packages that a port/package needs for it to be built and/or
run.
Port Makefiles refer to this type of ports/packages using the
.Dv BUILD_DEPENDS
and
.Dv RUN_DEPENDS
macros, respectively.
.It dependent
Ports/packages that need this port/package.
.El
.Sh OPTIONS
The following command line arguments are supported:
.Pp
.Bl -tag -width "--updateindex" -compact
.It Ar pkgname_glob
Specify one of these: a full pkgname, a pkgname without version, a
shell glob pattern in which you can use wildcards
.Sq * ,
.Sq \&? ,
and
.Sq [..]
(e.g.
.Dq Ar zsh ,
.Dq Ar gnome* ,
.Dq Ar bash-2* ,
etc.), or an extended regular expression preceded by a colon
.Sq \&: .
.Pp
.Nm ports_glob
lists ports which pkgnames match the pattern.
.Pp
.It Ar portorigin_glob
Specify a shell glob pattern. (e.g.
.Dq Ar archivers/p5-* ) ,
or an extended regular expression preceded by a colon
.Sq \&: .
.Pp
.Nm ports_glob
lists ports that match the pattern.
.Pp
Do not forget to include a slash if you want
.Nm ports_glob
to treat a pattern as a portorigin glob.
.Pp
.It Fl h
.It Fl -help
Show help and exit.
.Pp
.It Fl f
.It Fl -force
Force to update database regardless of timestamps.
.Pp
.It Fl F
.It Fl -fetchindex
Fetch the ports index file called
.Pa INDEX
from the official site.
.Pp
.It Fl M
.It Fl -master-recursive
List all master ports of the given ports as well.
It also lists the
master ports of all those required by the given ports if
.Fl R
is specified.
.Pp
.It Fl q
.It Fl -noconfig
Do not read the configuration file -
.Pa $PREFIX/etc/pkgtools.conf .
.Pp
.It Fl r
.It Fl -recursive
List all those ports depending on the given ports as well.
.Pp
.It Fl R
.It Fl -upward-recursive
List all those ports required by the given ports as well.
.Pp
.It Fl u
.It Fl -update
Update or create the ports database file
.Pa INDEX.db
from the ports index file.
It is in binary form and meant to be fast
to search for information.
.Pp
Note: the ports database file is automatically updated if it is not
up-to-date when looked up, so manual updating is not mandatory.
.Pp
.It Fl U
.It Fl -updateindex
Update or create the
.Pa INDEX
file by running
.Dq Nm make index .
If you define special macros in
.Pa /etc/make.conf
and the dependency of some ports are changed, you should create
.Pa INDEX
by yourself by using this option.
Otherwise, using
.Fl F
option is recommended since it is much faster.
.Pp
.It Fl x Ar portorigin_glob
.It Fl -exclude Ar portorigin_glob
Exclude ports matching the specified glob pattern from the list.
.El
.Sh ENVIRONMENT
.Bl -tag -width "PKGTOOLS_CONF" -compact
.It Ev PORTSDIR
Alternative location for the ports tree.
Default is
.Dq Pa /usr/ports .
.Pp
.It Ev PORTS_INDEX
Alternative location for the ports INDEX file.
Default is
.Dq Pa $PORTSDIR/INDEX .
.Pp
It is recommendable that you set this variable to something other than
the default to avoid conflict with CVS, CVSup, or CTM.
.Pp
.It Ev PORTS_DBDIR
Alternative location for the ports database file.
The database file
will be located in the first writable directory in the following ones:
.Ev $PORTS_DBDIR ,
.Ev $PORTSDIR ,
.Pa /usr/ports ,
.Ev $PKG_DBDIR ,
.Pa /var/db/pkg ,
.Ev $TMPDIR ,
.Pa /var/tmp ,
and
.Pa /tmp .
.Pp
.It Ev PKGTOOLS_CONF
Configuration file for the pkgtools suite.
Default is
.Dq Pa $PREFIX/etc/pkgtools.conf .
.El
.Sh FILES
.Bl -tag -width "$PREFIX/etc/pkgtools.conf" -compact
.It Pa /var/db/pkg
Default location of the installed package database.
.Pp
.It Pa /usr/ports
Default location of the ports tree.
.Pp
.It Pa INDEX
Ports index file, located right under the ports tree.
.Pp
.It Pa INDEX.db
Ports database file, located in
.Pa $PORTS_DBDIR .
.Pp
.It Pa $PREFIX/etc/pkgtools.conf
Default location of the pkgtools configuration file.
.El
.Sh SEE ALSO
.Xr pkgdb 1 ,
.Xr portversion 1 ,
.Xr pkgtools.conf 5 ,
.Xr ports 7
.Sh AUTHORS
.An Akinori MUSHA Aq knu@iDaemons.org
.Sh BUGS
.Fl S / Fl -slave-recursive
is not implemented yet.
.Pp
.Fl M
is very slow due to the limitation of
.Xr make 1
and
.Xr ports 7 .


NEWS
                           ======================
                            NEWS for portupgrade
                           ======================

------------------------------------------------------------------------

portupgrade 2.4.9.3 (released 2011-08-23):

* Bugfixes.

portupgrade 2.4.9.2 (released 2011-08-22):

* Bugfixes.

portupgrade 2.4.9.1 (released 2011-08-22):

* Ruby 1.9 related bugfixes.

portupgrade 2.4.9 (released 2011-08-18):

* Ruby 1.9 compatibility.

portupgrade 2.4.8 (released 2010-11-23):

* Bugfixes.

portupgrade 2.4.7 (released 2010-11-08):

* Bugfixes.

portupgrade 2.4.6 (released 2008-07-11):

* Fix man pages install.

portupgrade 2.4.5 (released 2008-07-10):

* [portversion, portversion.1] Document new options.

* [pkgtools.rb] Do not touch +CONTENTS file if it's not necessary.

* [pkgmisc.rb] Remove temporary dir with `rm -r'. It could be inner dirs there.

* [pkgdu] A new utility to display disk usage of installed packages.

portupgrade 2.4.4 (released 2008-07-01):

* [pkgtools.rb] Close +CONTENTS file. It made a few problems before.

* [portupgrade.1] Some leftovers removed. Fix .Pp tag.

* [portupgrade] Fix a failure when -PP specified for some ports.

* [portversion] Add -F to show a package full name (with a version number)
  -o to show an origin of a package and -q for quiet output.

* [portupgrade, pkgmisc.rb] Fix a bug in make arguments quoting. If you
  had an argument with spaces (e.g. WITH_MODULES="module1 module2"), it
  was parsed wrong.

* [portupgrade] Allow upgrade dependencies if -R or -r options specified
  but main port is up-to-date.

* [pkgdb.rb, portsdb.rb] If DB can't be updated, remove it and try again.

* [portupgrade] setproctitle with a port name when 'make config' runs.

* [portupgrade] If 'make config' fails, show a warning and ignore it.

* [pkgdb] Do not show 'You do not own pkgdb dir' for root.

portupgrade 2.4.3 (released 2008-02-11):

* [portupgrade] Fix dependencies list after we gather it. So we will not
  try to upgrade dependecies when a specified port is up-to-date.

* [portsclean] Fix a typo in a condition. Now it's work again.

* [portupgrade] Change a semantic of -q option. Now it means 'quiet'.
  Don't show a message when -N specified and there is already installed
  package. (For Peter Hofer. DesktopBSD).

portupgrade 2.4.2 (released 2008-02-04):

* A few bugs was fixed.

portupgrade 2.4.1 (released 2008-01-29):

* Bugs fix releasse.

* [portupgrade] Throw RecursiveDependencyError exception instead of
  plain ruby error.

portupgrade 2.4.0 (released 2008-01-26):

* [portupgrade] All dependencies run under portupgrade control now.

* [portupgrade] -c and -C options was changed. Now portupgrade
  run `make config-conditional` and `make config` accordingly.

* [pkgdb] Bug fix: remove +REQUIRED_BY file if there are no ports
  require this one.

portupgrade 2.3.2 (released 2008-01-13):

* Bugs fix release

portupgrade 2.3.1 (released 2007-07-03):

* [portupgrade] [pkgbd] Many bug fixes related to the last xorg update
  to 7.2.0

portupgrade 2.3.0 (released 2007-03-01):

* [portupgrade] At last all dependencies are tracked by portupgrade.

portupgrade 2.2.6 (released 2007-02-27):

* [pkgdb] Fix -O option parsing.

* [portsdb.db] Fix and update language specific categories list.

* [pkgdb] When run as pkg_which, don't check file existence. It allows to check
  against accidently removed files. (Asked by skv@).

* [pkgdb] Fix a bug when only origins worked in ALT_PKGDEP. pkgname globs
  are very helpful there. (e.g. 'mysql*-server' => 'mysql40-server').

portupgrade 2.2.5 (released 2007-02-26):

* [pkgdb] Fix issues with ghostscript-afpl vs. ghostscript-gnu,
  apache13 vs. apache13+mod-ssl, etc. when -F or -L. The ports must be
  described in ALT_PKGDEP section of pkgtools.conf.

* [pkgdb] Do not make broken pkgdb.db if no ports installed. nanobsd needs it.
  (Asked by Nick Hibma <n_hibma@...>)

* [pkgdb.rb] Speed-up autofix() - add -O option for pkgdb -aF execution.

* [pkgtools.conf] Describe PKG_SUFX variable. (Submitted by Pavel Gubin).

portupgrade 2.2.4 (released 2007-02-23):

* [pkgdb] Fix a bug when some dependencies could lost.

* [pkgdb] -L to allow users fix dependencies those was lost with the bug
  mentioned above.

* [pkgdb] -i turns interactive mode on.

* [pkgdb] Add -O option to turn off dependencies check instead of
  -d option, that turned it on.

portupgrade 2.2.3 (released 2007-02-16, never released in ports):

* Fixes reflected last changes in the ports tree.

* Hide pkgdb -F smartness under -d switch. It makes life harder when
  there are a lot of ports with a lot of depends. (Asked by marcus@).

portupgrade 2.2.2 (released 2006-11-18):

* Add UPGRADE_PORT_VER environment variable. Discussed with DougB@ and
  skv@.

* Add description of UPGRADE_* environment variables in a portupgrade
  man page.

* Add checking size of a lock file before trying read it. It protects us
  from a ruby error if the file is empty by some reason.
  (Reported by: Lowell Gilbert <freebsd-ports-local@be-well.ilk.org>)

portupgrade 2.2.1 (released 2006-11-12):

* Raise an error when MOVED file has a wrong format instead of a weird
  ruby error. (Reported by kris@FreeBSD.org).

* Fix a bug when -P always treated as -PP.

* Change PORT_UPGRADE environment variable with UPGRADE_PORT (contains
  a package name for updating port) and UPGRADE_TOOL=portupgrade.

lib/pkgtools/portsdb.rb
# vim: set sts=2 sw=2 ts=8 et:
#
# Copyright (c) 2001-2004 Akinori MUSHA <knu@iDaemons.org>
# Copyright (c) 2006-2008 Sergey Matveychuk <sem@FreeBSD.org>
# Copyright (c) 2009-2012 Stanislav Sedov <stas@FreeBSD.org>
# Copyright (c) 2012 Bryan Drewery <bdrewery@FreeBSD.org>
#
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#

require 'singleton'
require 'tempfile'
require 'pkgtools/pkgmisc'
require 'pkgtools/pkgdbtools'

class PortsDB
  include Singleton
  include Enumerable
  include PkgDBTools

  DB_VERSION = [:FreeBSD, 3]

  LANGUAGE_SPECIFIC_CATEGORIES = {
    "arabic"		=> "ar-",
    "chinese"		=> "zh-",
    "french"		=> "fr-",
    "german"		=> "de-",
    "hebrew"		=> "iw-",
    "hungarian"		=> "hu-",
    "japanese"		=> "ja-",
    "korean"		=> "ko-",
    "polish"		=> "pl-",
    "portuguese"	=> "pt-",
    "russian"		=> "ru-",
    "ukrainian"		=> "uk-",
    "vietnamese"	=> "vi-",
  }

  MY_PORT = 'ports-mgmt/portupgrade'

  LOCK_FILE = '/var/run/portsdb.lock'

  attr_accessor :ignore_categories, :extra_categories, :moved

  class IndexFileError < StandardError
#    def message
#      "index file error"
#    end
  end

  class IndexFileFetchError< IndexFileError
#    def message
#      "index file fetch error"
#    end
  end

  class DBError < StandardError
#    def message
#      "database file error"
#    end
  end

  class MOVEDError < StandardError
#    def message
#      "MOVED file error"
#    end
  end

  class MovedElement
    attr_reader :to, :date, :why, :seq
    def initialize(to, date, why, seq)
      @to = to
      @date = date
      @why = why
      @seq = seq
    end
  end

  class Moved
    MOVED_FILE = 'MOVED'

    def initialize(ports_dir)
      @moved = Hash.new
      @seq = 0
      fill(File.join(ports_dir, MOVED_FILE))
    end

    def fill(moved_file)
      if File.exist?(moved_file)
	File.open(moved_file) do |f|
	  f.each do |line|
	    next if /^[#[:space:]]/ =~ line

	    moved_from, moved_to, date, why = line.chomp.split('|')

	    if moved_from.nil? || moved_to.nil? || date.nil? || why.nil?
	      raise MOVEDError, "MOVED file format error"
	    end

	    moved_to.empty? and moved_to = nil

	    @moved[moved_from] = MovedElement.new(moved_to, date, why, @seq)
	    @seq += 1
	  end
	end
      end
    end

    def trace(port)
      t = []
      me = port

      while true
	if moved = @moved[me]
	  t << moved if t.empty? or t.last.seq < moved.seq
	  if me.nil? or t.map{|p| p.to}.include?(me)
	    break
	  else
	    me = moved.to
	  end
	else
	  break
	end
      end

      if t.empty?
	nil
      else
	t
      end
    end
  end

  def PortsDB.finalizer
    Proc.new {
      PkgDBTools.remove_lock(LOCK_FILE)
    }
  end

  def setup(alt_db_dir = nil, alt_ports_dir = nil, alt_db_driver = nil)
    @db = nil
    @lock_file = Process.euid == 0 ? LOCK_FILE : nil
    @db_version = DB_VERSION
    ObjectSpace.define_finalizer(self, PortsDB.finalizer)
    set_ports_dir(alt_ports_dir)
    set_db_dir(alt_db_dir)
    set_db_driver(alt_db_driver)

    @categories = nil
    @virtual_categories = nil
    @ignore_categories = []
    @extra_categories = []
    @origins = nil
    @pkgnames = nil
    @origins_by_categories = {}
    @ports = {}
    @localbase = nil
    @x11base = nil
    @pkg_sufx = nil
    @moved = Moved.new(ports_dir)

    self
  end

  def make_var(var, dir = ports_dir())
    all_vars = var.is_a?(Array) ? var : [var]
    vars = all_vars.join(' -V ')
    file = dir == ports_dir() ? "-f Mk/bsd.port.mk" : ""
    results = `cd #{dir} && make #{file} -V #{vars}`.lines.map { |val|
      val.strip!
      if val.empty?
        nil
      else
        val
      end
    }
    # Return array if requested, or single value if not
    var.is_a?(Array) ? results : results.first
  end

  def ports_dir()
    unless @ports_dir
      set_ports_dir(nil)	# initialize with the default value
    end

    @ports_dir
  end

  def ports_dir=(new_ports_dir)
    @abs_ports_dir = @index_file = @dist_dir = nil
    @alt_index_files = Array.new

    @ports_dir = new_ports_dir || ENV['PORTSDIR'] || '/usr/ports'
  end
  alias set_ports_dir ports_dir=

  def abs_ports_dir()
    unless @abs_ports_dir
      dir = ports_dir

      begin
	Dir.chdir(dir) {
	  @abs_ports_dir = Dir.pwd
	}
      rescue => e
	raise DBError, "Can't chdir to '#{dir}': #{e.message}"
      end
    end

    @abs_ports_dir
  end

  def index_file()
    unless @index_file
      indexdir, indexfile = make_var(['INDEXDIR', 'INDEXFILE'])
      @index_file = ENV['PORTS_INDEX'] || File.join(indexdir, indexfile || 'INDEX')
      @alt_index_files = config_value(:ALT_INDEX) || []
    end

    @index_file
  end

  def db_dir=(new_db_dir)
    @db_dir = new_db_dir || ENV['PORTS_DBDIR'] || ports_dir

    @db_filebase = File.join(@db_dir, File.basename(index_file()))
    @db_file = @db_filebase + '.db'

    close_db

    @db_dir
  end
  alias set_db_dir db_dir=

  def db_dir_list()
    [
      db_dir,
      ports_dir,
      PkgDB.instance.db_dir,
      ENV['TMPDIR'],
      '/var/tmp',
      '/tmp'
    ].compact
  end

  def my_port
    MY_PORT
  end

  def my_portdir
    portdir(MY_PORT)
  end

  def localbase
    @localbase ||= make_var('LOCALBASE', my_portdir) || '/usr/local'
  end

  def x11base
    @x11base ||= make_var('X11BASE', my_portdir) || '/usr/X11R6'
  end

  def pkg_sufx
    @pkg_sufx ||= pkg_sufx!
  end

  def pkg_sufx!
    make_var('PKG_SUFX', my_portdir) || ENV['PKG_SUFX'] || '.tbz'
  end

  def dist_dir()
    @dist_dir ||= make_var('DISTDIR') || portdir('distfiles')
  end

  def join(category, port)
    File.join(category, port)
  end

  def split(origin)
    if %r"^([^./A-Z][^/]*)/([^./][^/]*)$" =~ path
      return $1, $2
    end

    nil
  end

  def strip(path, existing_only = false)
    # handle sequences of /'s (tr_s is not multibyte-aware, hence gsub)
    path = path.gsub(%r"//+", '/')

    %r"^(?:(/.+)/)?([^./][^/]*/[^./][^/]*)/?$" =~ path or return nil

    dir = $1
    port = $2

    if dir && dir != ports_dir && dir != abs_ports_dir
      return nil
    end

    if existing_only && !exist?(port)
      return nil
    end

    port
  end

  def portdir(port)
    File.join(ports_dir, port)
  end

  def subdirs(dir)
    %x"fgrep SUBDIR #{dir}/Makefile | sed -e 's/SUBDIR +=//'
       2> /dev/null".split.select { |i|
      File.directory?(File.join(dir, i))
    }.sort
  end

  def categories
    open_db if @categories.nil?

    @categories
  end

  def real_categories!
    subdirs(ports_dir)
  end

  def categories!
    customize_categories(real_categories!)
  end

  def customize_categories(cats)
    ((cats | @extra_categories) - @ignore_categories).sort
  end

  def category?(category)
    @categories.qinclude?(category)
  end

  def virtual_categories
    open_db if @virtual_categories.nil?

    @virtual_categories
  end

  def virtual_category?(category)
    @virtual_categories.qinclude?(category)
  end

  def ignore_category?(category)
    @ignore_categories.qinclude?(category)
  end

  def update(fetch = false)
    if fetch
      STDERR.print "Fetching the ports index ... "
    else
      STDERR.print "Updating the ports index ... "
    end

    STDERR.flush

    t = Tempfile.new('INDEX')
    t.close
    tmp = t.path

    if File.exist?(index_file)
      if !File.writable?(index_file)
	STDERR.puts "index file #{index_file} not writable!"
	raise IndexFileError, "index generation error"
      end
    else
      dir = File.dirname(index_file)

      if !File.writable?(dir)
	STDERR.puts"index file directory #{dir} not writable!"
	raise IndexFileError, "index generation error"
      end
    end

    if fetch
      system "cd #{abs_ports_dir} && make fetchindex && cp #{index_file} #{tmp}"
    else
      system "cd #{abs_ports_dir} && make INDEXFILE=INDEX.tmp index && mv INDEX.tmp #{tmp}"
    end

    if File.zero?(tmp)
      if fetch
        STDERR.puts 'failed to fetch INDEX!'
        raise IndexFileFetchError, "index fetch error"
      else
        STDERR.puts 'failed to generate INDEX!'
        raise IndexFileError, "index generation error"
      end
    end

    begin
      File.chmod(0644, tmp)
    rescue => e
      STDERR.puts e.message
      raise IndexFileError, "index chmod error"
    end

    if not system('/bin/mv', '-f', tmp, index_file)
      STDERR.puts 'failed to overwrite #{index_file}!"'
      raise IndexFileError, "index overwrite error"
    end

    STDERR.puts "done"

    @categories = nil
    @virtual_categories = nil
    @origins = nil
    @pkgnames = nil
    @origins_by_categories = {}
    @ports = {}

    close_db
  end

  def open_db
    @db and return @db

    update_db

    retried = false

    begin
      open_db_for_read!

      check_db_version or raise TypeError, 'database version mismatch/bump detected'

      s = @db[':categories']
      s.is_a?(String) or raise TypeError, "missing key: categories"
      @categories = s.split

      s = @db[':virtual_categories']
      s.is_a?(String) or raise TypeError, "missing key: virtual_categories"
      @virtual_categories = s.split

      s = @db[':origins']
      s.is_a?(String) or raise TypeError, "missing key: origins"
      @origins = s.split

      s = @db[':pkgnames']
      s.is_a?(String) or raise TypeError, "missing key: pkgnames"
      @pkgnames = s.split.map { |n| PkgInfo.new(n) }

      @origins_by_categories = {}
      (@categories + @virtual_categories).each do |c|
	s = @db['?' + c] and @origins_by_categories[c] = s.split
      end
    rescue => e
      if retried
	raise DBError, "#{e.message}: Cannot read the portsdb!"
      end

      STDERR.print "[#{e.message}] "
      update_db(true)

      retried = true
      retry
    end

    @ports = {}

    @db
  rescue => e
    STDERR.puts e.message
    raise DBError, 'database file error'
  end

  def date_index
    latest_mtime = File.mtime(index_file) rescue nil
    @alt_index_files.each do |f|
      mt = File.mtime(f)
      latest_mtime = mt if mt > latest_mtime
    end
    latest_mtime
  end

  def date_db
    File.mtime(@db_file) rescue nil
  end

  def up_to_date?
    d1 = date_db() and d2 = date_index() and d1 >= d2
  end

  def select_db_dir(force = false)
    return db_dir if File.writable?(db_dir)

    db_dir_list.each do |dir|
      set_db_dir(dir)

      !force && up_to_date? and return dir

      File.writable?(dir) and return dir
    end

    nil
  end

  def update_db(force = false)
    if not File.exist?(index_file)
      begin
        update(true)
      rescue IndexFileFetchError
        update(false)
      end
    end

    !force && up_to_date? and return false

    close_db

    select_db_dir(force) or raise "No directory available for portsdb!"

    prev_sync = STDERR.sync
    STDERR.sync = true

    STDERR.printf "[Updating the portsdb <format:%s> in %s ... ", db_driver, db_dir

    nports = `wc -l #{index_file}`.to_i
    @alt_index_files.each do |f|
      nports += `wc -l #{f}`.to_i
    end

    STDERR.printf "- %d port entries found ", nports

    i = -1

    @origins = []
    @pkgnames = []

    try_again = false
    begin
      open_db_for_rebuild!

      index_files = shelljoin(index_file) + ' '
      index_files.concat(@alt_index_files.join(' '))

      open("| sort #{index_files}", 'r:utf-8') do |f|
	f.each_with_index do |line, i|
	  lineno = i + 1

	  if lineno % 100 == 0
	    if lineno % 1000 == 0
	      STDERR.print lineno
	    else
	      STDERR.putc(?.)
	    end
	  end

	  begin
	    port_info = PortInfo.new(line)

	    next if ignore_category?(port_info.category)

	    origin = port_info.origin
	    pkgname = port_info.pkgname

	    port_info.categories.each do |category|
	      if @origins_by_categories.key?(category)
		@origins_by_categories[category] << origin
	      else
		@origins_by_categories[category] = [origin]
	      end
	    end
	    
	    @ignore_categories.each do |category|
	      @origins_by_categories.delete(category)
	    end

	    @origins << origin
	    @pkgnames << pkgname

	    @db[origin] = port_info
	    @db[pkgname.to_s] = origin
	  rescue => e
	    STDERR.puts index_file + ":#{lineno}:#{e.message}"
	  end
	end
      end

      STDERR.print ' '

      real_categories = real_categories! | @extra_categories
      all_categories = @origins_by_categories.keys

      @categories = (real_categories - @ignore_categories).sort
      @virtual_categories = (all_categories - real_categories).sort

      @db[':categories'] = @categories.join(' ')
      STDERR.putc(?.)
      @db[':virtual_categories'] = @virtual_categories.join(' ')
      STDERR.putc(?.)
      @db[':origins'] = @origins.join(' ')
      STDERR.putc(?.)
      @db[':pkgnames'] = @pkgnames.map { |n| n.to_s }.join(' ')
      STDERR.putc(?.)
      all_categories.each do |c|
	@db['?' + c] = @origins_by_categories[c].join(' ')
      end
      STDERR.putc(?.)
      @db[':db_version'] = Marshal.dump(DB_VERSION)
    rescue => e
      if File.exist?(@db_file)
	begin
	  STDERR.puts " error] Remove and try again."
	  File.unlink(@db_file)
	  try_again = true
	rescue => e
	  raise DBError, "#{e.message}: Cannot update the portsdb! (#{@db_file})]"
	end
      else
	raise DBError, "#{e.message}: Cannot update the portsdb! (#{@db_file})]"
      end
    ensure
      close_db
    end

    if try_again
      update_db(force)
    else
      STDERR.puts " done]"
      STDERR.sync = prev_sync
      true
    end

  end

  def port(key)
    key.is_a?(PortInfo) and return key

    @ports.key?(key) and return @ports[key]

    open_db

    if key.include?('/')
      val = @db[key]
    elsif val = @db[key]
      return port(val)
    end

    @ports[key] = if val then PortInfo.new(val) else nil end
  end
  alias [] port

  def ports(keys)
    keys.map { port(key) }
  end

  alias indices ports

  def origin(key)
    if p = port(key)
      p.origin
    else
      nil
    end
  end

  def origins(category = nil)
    open_db

    if category
      @origins_by_categories[category]
    else
      @origins
    end
  end

  def origins!(category = nil)
    if category
      # only lists the ports which primary category is the given category
      subdirs(portdir(category)).map { |i|
	File.join(category, i)
      }
    else
      list = []

      categories!.each do |i|
	list.concat(origins!(i))
      end

      list
    end
  end

  def each(category = nil)
    ports = origins(category) or return nil

    ports.each { |key|
      yield(@db[key])
    }
  end

  def each_category
    categories.each { |key|
      yield(key)
    }
  end

  def each_origin(category = nil)
    ports = origins(category) or return nil

    ports.each { |key|
      yield(key)
    }
  end

  def each_origin!(category = nil, &block)
    if category
      # only lists the ports which primary category is the given category
      subdirs(portdir(category)).each do |i|
	block.call(File.join(category, i))
      end
    else
      categories!.each do |i|
	each_origin!(i, &block)
      end
    end
  end

  def each_pkgname
    open_db

    @pkgnames.each { |key|
      yield(key)
    }
  end

  def glob(pattern = '*')
    list = []
    pkg = nil

    open_db

    case pattern
    when Regexp
      is_port = pattern.source.include?('/')
    else
      if /^[<>]/ =~ pattern
	raise "Invalid glob pattern: #{pattern}"
      end

      is_port = pattern.include?('/')

      # shortcut
      if portinfo = port(pattern)
	if block_given?
	  yield(portinfo)
	  return nil
	else
	  return [portinfo]
	end
      end
    end

    if is_port
      @origins.each do |origin|
	case pattern
	when Regexp
	  next if pattern !~ origin
	else
	  next if not File.fnmatch?(pattern, origin, File::FNM_PATHNAME)
	end

	if portinfo = port(origin)
	  if block_given?
	    yield(portinfo)
	  else
	    list.push(portinfo)
	  end
	end
      end
    else
      @pkgnames.each do |pkgname|
	next if not pkgname.match?(pattern)

	if portinfo = port(pkgname.to_s)
	  if block_given?
	    yield(portinfo)
	  else
	    list.push(portinfo)
	  end
	end
      end
    end

    if block_given?
      nil
    else
      list
    end
  rescue => e
    STDERR.puts e.message

    if block_given?
      return nil
    else
      return []
    end
  end

  def exist?(port, quick = false)
    return if %r"^[^/]+/[^/]+$" !~ port

    dir = portdir(port)

    return false if not File.file?(File.join(dir, 'Makefile'))

    return true if quick

    make_var('PKGNAME', dir) || false
  end

  def all_depends_list!(origin, before_args = nil, after_args = nil)
    `cd #{$portsdb.portdir(origin)} && #{before_args || ''} make #{after_args || ''} all-depends-list`.lines.map { |line|
      strip(line.chomp, true)
    }.compact
  end

  def all_depends_list(origin, before_args = nil, after_args = nil)
    if !before_args && !after_args && i = port(origin)
      i.all_depends.map { |n| origin(n) }.compact
    else
      all_depends_list!(origin, before_args, after_args)
    end
  end

  def masters(port)
    dir = portdir(port)

    ports = []

    `cd #{dir} ; make -dd -n 2>&1`.each do |line|
      if /^Searching for .*\.\.\.Caching .* for (\S+)/ =~ line.chomp
	path = File.expand_path($1)

	if (path.sub!(%r"^#{Regexp.quote(ports_dir)}/", '') ||
	    path.sub!(%r"^#{Regexp.quote(abs_ports_dir)}/", '')) &&
	    %r"^([^/]+/[^/]+)" =~ path
	  x = $1

	  ports << x if exist?(x) && !ports.include?(x)
	end
      end
    end

    ports.delete(port)

    ports
  end

  def latest_link(port)
    dir = portdir(port)

    flag, name = make_var(['NO_LATEST_LINK', 'LATEST_LINK'], dir)

    if flag
      nil
    else
      name
    end
  end

  def sort(ports)
    tsort = TSort.new

    ports.each do |p|
      portinfo = port(p)

      portinfo or next

      o = portinfo.origin

      deps = all_depends_list(o)	# XXX

      tsort.add(o, *deps)
    end

    tsort.tsort! & ports
  end

  def sort!(ports)
    ports.replace(sort(ports))
  end

  def recurse(portinfo, recurse_down = false, recurse_up = false)
    if not portinfo.is_a?(PortInfo)
      portinfo = port(portinfo)
    end

    list = []

    portinfo or return list

    if recurse_up
      portinfo.all_depends.map do |name|
	i = port(name)

	list << i if i
      end
    end

    list << portinfo

    if recurse_down
      # slow!
      pkgname = portinfo.pkgname.fullname

      glob do |i|
	list << i if i.all_depends.include?(pkgname)
      end
    end

    list
  end
end
